### PR TITLE
CI: clean the CI workflow by using Dockerfile

### PR DESCRIPTION
**Pull request by Zheaoli**
## Description

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by Officeyutong** [comment]
It would be better to lock the version of images. Unexpected new version may lead to incorrect package names

**Comment by Officeyutong** [comment]
Is it necessary to move the setup section into separate dockerfile? I think it's not necessary. Keep them in the workflow file making it more easier to update CI

But if be publish the docker image to ghcr would save more CI time. But this will make it more complex to update CI. I think it should be discussed

**Comment by Officeyutong** [comment]
Besides, I'm not sure if the "container" section in workflow syntax could use dockerfile(In my knowledge, a dockerfile is required to be built into an image before using)

**Comment by Zheaoli** [comment]
> Besides, I'm not sure if the "container" section in workflow syntax could use dockerfile(In my knowledge, a dockerfile is required to be built into an image before using)

Yes, this PR is just for test, I'm trying to find a way to clean the CI

**Comment by Officeyutong** [comment]
> > Besides, I'm not sure if the "container" section in workflow syntax could use dockerfile(In my knowledge, a dockerfile is required to be built into an image before using)
> 
> Yes, this PR is just for test, I'm trying to find a way to clean the CI

You may publish the docker images to ghcr and utilize them

**Comment by Officeyutong** [comment]
> > Besides, I'm not sure if the "container" section in workflow syntax could use dockerfile(In my knowledge, a dockerfile is required to be built into an image before using)
> 
> Yes, this PR is just for test, I'm trying to find a way to clean the CI

Testing purpose might be done in your own fork. Using PR will cost the CI amount of organization eunomia-bpf

**Comment by Officeyutong** [comment]
> > > Besides, I'm not sure if the "container" section in workflow syntax could use dockerfile(In my knowledge, a dockerfile is required to be built into an image before using)
> 
> > 
> 
> > Yes, this PR is just for test, I'm trying to find a way to clean the CI
> 
> 
> 
> You may publish the docker images to ghcr and utilize them

Note that we need a suitable way to update and publish these images when changes were made

**Comment by Officeyutong** [comment]
Would you like to get this PR finally merged? If not, I think it's better to test CI in your own fork. CI time amount is limited in organization account

**Comment by yunwei37** [comment]
> Is it necessary to move the setup section into separate dockerfile? I think it's not necessary. Keep them in the workflow file making it more easier to update CI
> 
> But if we publish the docker image to ghcr, more CI time would be saved. But this will make it more complex to update CI. I think it should be discussed

I think it could be a possible way to use docker images in the ci?

There are a lot of repeat code, make it hard to maintain the ci scripts.

Maybe another solution is writing some bash scripts to setup the environments for ci test.

File: .github/workflows/test-aot-cli.yml [hunk 1]
```diff
@@ -5,41 +5,26 @@ on:
...
-        runs-on-container:
-          - ubuntu:22.04
-          - fedora:39
+        container:
+          - ubuntu-2204
+          - fedora-39
     runs-on: ubuntu-latest
     container: 
-      image: ${{matrix.runs-on-container}}
+      image: "manjusakalza/bpftime-base-image:${{matrix.container}}"
```
Comment (Officeyutong) [inline .github/workflows/test-aot-cli.yml:20]: I think these images should be owned by account eunomia-bpf

**Review by Officeyutong — CHANGES_REQUESTED** [review]

**Comment by Zheaoli** [comment]
For now, I wish this PR to be merged. I choose to use a pre-built Docker image in the CI process to avoid duplicated environment-prepared code and unnecessary conditions in the workflow. 

In the future, if we need add more test environment, we just need add a new image to adapt,  so we can keep the core test in workflow Simple and clean

File: .github/workflows/test-aot-cli.yml (hunk 1 above)
Comment (Zheaoli) [reply .github/workflows/test-aot-cli.yml:20]: Yes, I will update the image and add a new workflow to build the image when we reach consensus

File: .github/workflows/test-runtime.yml [hunk 2]
```diff
@@ -5,40 +5,28 @@ on:
...
   pull_request: 
     branches: "master"
-
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name }}
+  cancel-in-progress: true
 jobs:
   build:
-    runs-on: ubuntu-22.04
+    runs-on: "ubuntu-latest"
```
[the remark this answers is not in the document]
Comment (Zheaoli) [reply .github/workflows/test-runtime.yml:14]: Got it

File: .github/workflows/test-aot-cli.yml (hunk 1 above)
Comment (Officeyutong) [reply .github/workflows/test-aot-cli.yml:20]: > Yes, I will update the image and add a new workflow to build the image when we reach consensus

What about moving these images to eunomia-bpf now? Do you have permission to publish image?

**Review by yunwei37 — APPROVED** [review]

**Comment by yunwei37** [comment]
Thanks! Happy new year!

**Pull request merged**